libvast/src/system/node.cpp

```diff
@@ -27,192 +29,207 @@
+  // Auto-generate label if none given.
+  string label;
+  if (auto label_ptr = caf::get_if<string>(&options, "global.label")) {
+    label = *label_ptr;
+  } else {
+    label = comp_name;
+    const char* multi_instance[] = {"importer", "exporter", "source", "sink"};
+    if (std::count(begin(multi_instance), end(multi_instance), label) != 0) {
+      // Create a new label and update our counter in the map.
+      auto n = ++this_node->state.labels[label];
+      label += '-';
+      label += std::to_string(n);
+      VAST_DEBUG(this_node, "auto-generated new label:", label);
     }
   }
-  if (!fun) {
-    rp.deliver(make_error(ec::unspecified, "invalid spawn component"));
-    return;
-  }
-  // Parse arguments.
-  auto component_args = args.take_right(args.size() - i - 1);
-  auto spawn_args = args.take(i);
-  std::string label;
-  auto r = spawn_args.extract_opts({
-    {"label,l", "a unique label for the component", label}
-  });
-  if (!r.remainder.empty()) {
-    auto invalid = r.remainder.get_as<std::string>(0);
-    rp.deliver(make_error(ec::syntax_error, "invalid syntax", invalid));
-  }
-  if (!r.error.empty())
-    rp.deliver(make_error(ec::syntax_error, std::move(r.error)));
-  // Register the component with the track.
-  self->request(self->state.tracker, infinite, get_atom::value).then(
-    [=](registry& reg) mutable {
-      VAST_ASSERT(reg.components.count(self->state.name) > 0);
-      auto& local = reg.components[self->state.name];
-      // Check if we can spawn more than one instance of the given component.
-      for (auto c : {"metastore", "archive", "index", "profiler"})
-        if (c == component && local.count(component) > 0) {
-          rp.deliver(make_error(ec::unspecified, "component already exists"));
-          return;
-        }
-      // Auto-generate a label if none given.
-      if (label.empty()) {
-        label = component;
-        auto multi_instance = component == "importer"
-                           || component == "exporter"
-                           || component == "source"
-                           || component == "sink";
-        if (multi_instance) {
-          label += '-';
-          label += std::to_string(++self->state.labels[component]);
-          VAST_DEBUG(self, "auto-generated new label:", label);
-        }
-      }
-      // Dispatch spawn command.
-      auto opts = options{component_args, self->state.dir, label};
-      auto a = fun(opts);
-      if (!a) {
-        rp.deliver(a.error());
-        return;
-      }
-      self->request(self->state.tracker, infinite, put_atom::value, component,
-                    *a, label).then(
-        [=](ok_atom) mutable {
-          rp.deliver(*a);
-        },
-        [=](error& e) mutable {
-          rp.deliver(std::move(e));
-        }
-      );
+  // Spawn our new VAST component.
+  spawn_arguments args{cmd, st.dir, label, options, first, last};
+  caf::actor new_component;
+  if (auto spawn_res = spawn_component(cmd, args))
```

Comment:
How about `s/spawn_res/actor/` for more semantics?

Reply:
To me that adds confusion, because `spawn_component` doesn't return an actor. It maybe returns an actor. Since `maybe_actor` is already a type alias I picked this name to clearly indicate that the result of spawn might not be an actor.

Reply:
Okay, if `spawn_component` doesn't always return an actor, I'm fully onboard with something generic like result.